OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -984,7 +987,11 @@ void BuildUnit(string category, bool buildRandom)
 				ChooseRandomUnitToBuild(queue) :
 				ChooseUnitToBuild(queue);
 
-			if (unit != null && Info.UnitsToBuild.Any(u => u.Key == unit.Name))
+			if (unit != null
+				&& Info.UnitsToBuild.ContainsKey(unit.Name)
+				&& (!Info.UnitLimits.ContainsKey(unit.Name)
+					|| World.Actors.Count(a => a.Info.Name == unit.Name && a.Owner == Player) < Info.UnitLimits[unit.Name]))
```

Comment:
It feels like it'd be more sensible if ```ChooseRandomUnitToBuild``` and ```ChooseUnitToBuild``` took the unit limits into account (as well as ```Info.UnitsToBuild``` to be honest - might be worth seeing if you can move that too).